### DISCUSSION TITLE
Sending a can message on GMLAN

### DISCUSSION THREAD

**Discussion by ntrambitas12** [BrianHumlicek/J2534-Sharp#29] [Q&A]
Hi,
I stumbled across this library while looking for an interface to write a simple program to send messages on the GMLAN. My car is a 2014 Cadillac ELR. I have a packet that I want to try to send. I was told that the following message is correct to send (it commands the windows to roll down) but as I connect to the network and try to send, nothing happens and an exception is thrown. Am I not correctly setting up my connection to the GMLAN or formatting the CAN message correctly?

![image](https://user-images.githubusercontent.com/47403010/221418866-16e4c896-ba1c-435f-a09e-c8bd9596d5a6.png)
![image](https://user-images.githubusercontent.com/47403010/221418898-06742d12-d022-4247-9fe9-548373a76178.png)

Thank you for any help!

**Comment by BrianHumlicek** [comment]
I have never used GM LAN, but doing some reading, it appears its 29bit CAN among other things.  At the very least, you will probably need to use the 29bit flags on connection and on each message transmission.  This is assuming that using the CAN protocol is appropriate for GM LAN.  I suspect there might be another protocol definition that is intended to be used for that, but I don't know.

Brian

**Comment by BrianHumlicek** [comment]
Doing some reading, it appears you should be looking at one of the SW_CAN protocols.  SW_CAN_PS I suspect is the right one.  Also, I just added a baud definition for GM LAN.  It might take a minute to get nuget updated though.

**Reply by ntrambitas12** [reply]
Thank you so much, I'll have to experiment with it tonight and see if I can get it to work! Also, is there a specific way to format the CAN message (like specficying the header/module ID and/or how many bytes of data I'm sending) or do I just put the entire message in byte by byte. 

For example, I want to send the following message on the CAN network:

0x252 07 AE FE 80 00 00 00 00
(0x252 is Module ID/header)

Would the right command be:
  channel1.SendMessages(new[] { new Message(new byte[] { 0x02, 0x52, 0x07, 0xae, 0xfe, 0x80, 0x00, 0x00, 0x00, 0x00 }, TxFlag.CAN_29BIT_ID) }); 
? 

I want to make sure I'm formatting things correctly.

Thank you again

**Reply by BrianHumlicek** [reply]
Message format is going to be protocol specific, so I don't really have much input on what is proper for GM LAN.  I would recommend seeing if you can create a pass all filter and monitor all the bus traffic first.  That will probably give some insight into what the message should look like.

Brian

**Reply by ntrambitas12** [reply]
Hi Brian, I tried setting a pass all message filter and I've tried all the different protocols for single wire can and I keep running into the following exception:

![image](https://user-images.githubusercontent.com/47403010/222012820-e24bf524-a7cb-4679-9985-735241278807.png)

Is the GMLAN protocol a special protocol that this library does not support? I tried the normal High Speed ECU bus and it does work for sure since I can communicate and get a readout.

Thank you for all the help,
Nick

**Reply by BrianHumlicek** [reply]
I think you need to specify the pin you're communicating on using an IOCTL SetConfig J1962_PINS.  I'm not totally sure on the details, but I think that will get you going in the right direction.

The single wire CAN protocols were added with J2534-2, so, their implementation is a little clunky compared to the first class protocols like ISO15765.

Brian

**Reply by BrianHumlicek** [reply]
I'm don't know if you read it, but this discussion has some relevant info. #13

**Reply by ntrambitas12** [reply]
Thank you, 
That would make a lot of sense! GMLAN operates on Pin 1. I didn't think of manually setting the Pin. I will try it out and see if it works. Hopefully this will clear things up for anyone else trying to do this in the future

-Nick

**Reply by BrianHumlicek** [reply]
All of the protocols with the _PS suffix are ‘pin switched’ protocols, meaning the physical pins it originates from is not defined in the specification.  It needs to be configured with the connection.

This is j2534-2’s solution to adding more physical layer protocols than we have physical pins for.  It allows the user to pick and choose what physical layer configuration is needed.

Brian

**Reply by ntrambitas12** [reply]
Hey Brian, I've tried the change you mentioned but now it throws an exception as soon as I run the change pin command. It's "Error creating ComLogicalLink"

This is the code I have:
![image](https://user-images.githubusercontent.com/47403010/222624392-bdb4501e-f286-468c-8239-9e62482a94b7.png)

Any clues?

Thanks again for all your help,
Nick

**Reply by BrianHumlicek** [reply]
I'm not sure.  My only suggestion would be to remove the line that opens 'channel1' as maybe its conflicting with the GMLAN channel?  Alternately, its possible that your device doesn't support running that protocol on that pin.  Again, I'm not really sure, GMLAN isn't something I have experience with.

Brian

**Reply by ntrambitas12** [reply]
Hey Brian, 

I finally had some more time to poke around. So I have a feeling that I am just not setting the Pin correctly. I know for sure that my passThru is compatible with GMLAN since I used it before to reprogram GMLAN modules. GMLAN is on pin 1. I tried passing in 0x01 and just 1 and I get this exception. I tried commenting out the HSCAN channel and the same thing. I think its something related to the way Im trying to set the pins.

![image](https://user-images.githubusercontent.com/47403010/222980760-3c3705eb-4546-4afc-b1bb-42edcb497c44.png)

I feel like its so close to working

Thanks, 
Nick

**Reply by BrianHumlicek** [reply]
I’ve been there before myself.  For me it has been a combination of connect and txflags.  In the past I have used the debug mode built into Drewtechs API to sniff a known working implementation and then do what it did.  I’m not sure if tour device has anything similar.

If your device doesn’t have any thing similar, I think I may have seen a piece of software on GitHub that might do something similar.  I never used it so I don’t know for sure, but I’ll see if I can find it and link it.

Brian
